### PR TITLE
Skip indexing if a page has not changed

### LINKED ISSUE #2235: Search indexer performance penalty/glitches
**Affected version(s)**

(only tested with current master so far, probably at least 4.10)

**Description**

The search indexer locks and updates the search tables on every call even if nothing has changed (case A) with the exception of one scenario where the page was visited in preview mode first and therefore contains the wrong url (case B). When fixing this we should also make sure the lock only ever occurs after the checksum check/early return if possible.

**How to reproduce case A**
1. Make sure the search indexer runs in the FE
1. Open a page &rarr; will be indexed
1. Open the same page again without changing the content &rarr; `tl_search` and `tl_search_term` will be updated again<sup>*)</sup>

**How to reproduce case B**
1. Make sure the search indexer runs in the FE
1. Truncate the search tables
1. Open a page in the preview mode &rarr; the stored url will contain `preview.php/` 
1. Open the same page through the regular entry point &rarr; no change happens to the url or any associated record<sup>**)</sup>

--------
<sup>*)</sup> because there is no early return in the case of matching checksums and urls

<sup>**)</sup> because this first condition https://github.com/contao/contao/blob/fab8317d8f06000b8977987fb9dd71ee3d66415d/core-bundle/src/Resources/contao/library/Contao/Search.php#L188-L192 evaluates to `true` and the following to `false` resulting in a `return` in line 220.

### PR DESCRIPTION
**Pull request by ausi**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2235
| Docs PR or issue | -

Same as #2249 but for Contao 4.10.

### CODE REVIEW

File: core-bundle/src/Resources/contao/library/Contao/Search.php [hunk 1]
```diff
@@ -177,6 +177,17 @@ public static function indexPage($arrData)
 		// Calculate the checksum
 		$arrSet['checksum'] = md5($arrSet['text']);
 
+		$blnIndexExists = (bool) $objDatabase
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Search.php]: Why the cast? The if condition does not require type safety.

Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Search.php]: Should the variable name still be `$blnIndexExists` then? Or `$strIndexExists` instead?

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Search.php]: It actually does not matter. I guess we have been using `$bln` for any bool-ish value.

Will MySQL return `0` and `1` or `'0'` and `'1'`?

**Review by leofeyer — DISMISSED** [review]

File: core-bundle/src/Resources/contao/library/Contao/Search.php (hunk 1 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Search.php]: `'0'` and `'1'`

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Search.php]: I would still use `$bln` I guess.

Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Search.php]: Done in e3310bbc84f8d54dbbd8c8d956ba6a651a070aa7

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @ausi.

**Pull request merged**